File: litellm/manage-budget-key.py
#!/usr/bin/env python3
"""
manage-budget-key.py
--------------------
Manage a model-group-scoped LiteLLM virtual key with Claude-Code-style layered
budgets:

  - a weekly cap   (enforced on the user)
  - a window cap   (default 5h, enforced on the key)

The key blocks when EITHER limit is hit.

Modes:
  --create  (default)  Revoke any existing key for the user+group and mint a
                       NEW one. New sk-... value, window spend restarts at $0.
  --update             Change the budgets on the EXISTING key in place. Same
                       sk-... value; accumulated spend and reset window kept.
  --list               Show all managed budget keys as a table (budgets + spend;
                       the secret is never stored, so only a masked key is shown).
                       Optional --group filters to one group.

Model groups (--group, default "deepseek"):
  deepseek  -> models from the proxy whose id starts with "deepseek-"
  copilot   -> models whose id starts with "github_copilot/"
  (--models a,b,c overrides the list explicitly)

The group also names the key alias: "<group>-budget-<user_id>", which is how
one budgeted key per user per group is found and rotated without its plaintext.
Note: the weekly cap lives on the USER, so it aggregates across all of that
user's keys/groups; user-level model access is the union of all groups granted.

Usage:
  ./litellm/manage-budget-key.py [--create|--update] [--group G] [--models a,b]
                                 [user_id] [weekly_usd] [window_usd] [window]

Defaults: user_id=haitao  weekly_usd=20  window_usd=3  window=5h

Examples:
  ./litellm/manage-budget-key.py                            # deepseek key, $20/wk + $3/5h
  ./litellm/manage-budget-key.py --update haitao 20 5       # raise 5h cap, keep key+spend
  ./litellm/manage-budget-key.py --group copilot bob 30 5   # copilot key for bob

Requirements: proxy running (BASE, default http://localhost:4000);
LITELLM_MASTER_KEY set (sourced from .env at repo root if present).
"""

import argparse
import json
import os
import sys
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
GROUP_PREFIXES = {"deepseek": "deepseek-", "copilot": "github_copilot/"}
# Local plaintext store of minted secrets, keyed by alias. litellm only stores a
# hash, so this is the only way --list can show full keys. Git-ignored, chmod 600.
SECRETS_FILE = ROOT / ".litellm-budget-keys.json"


def load_secrets() -> dict:
    try:
        return json.loads(SECRETS_FILE.read_text())
    except Exception:
        return {}


def save_secret(alias: str, key: str) -> None:
    secrets = load_secrets()
    secrets[alias] = key
    SECRETS_FILE.write_text(json.dumps(secrets, indent=2))
    try:
        os.chmod(SECRETS_FILE, 0o600)
    except OSError:
        pass


def load_master_key() -> str:
    if not os.environ.get("LITELLM_MASTER_KEY"):
        env = ROOT / ".env"
        if env.is_file():
            for line in env.read_text().splitlines():
                line = line.strip()
                if line.startswith("LITELLM_MASTER_KEY="):
                    os.environ["LITELLM_MASTER_KEY"] = line.split("=", 1)[1]
                    break
    mk = os.environ.get("LITELLM_MASTER_KEY")
    if not mk:
        sys.exit("ERROR: LITELLM_MASTER_KEY is not set (export it or put it in .env).")
    return mk


def api(base: str, mk: str, method: str, path: str, body: dict | None = None) -> dict:
    req = urllib.request.Request(
        base + path,
        data=json.dumps(body).encode() if body is not None else None,
        headers={"Authorization": f"Bearer {mk}", "Content-Type": "application/json"},
        method=method,
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            return json.load(r)
    except urllib.error.HTTPError as e:
        detail = e.read().decode(errors="replace")[:400]
        raise SystemExit(f"ERROR: {method} {path} -> HTTP {e.code}: {detail}") from e
    except urllib.error.URLError as e:
        raise SystemExit(f"ERROR: cannot reach {base} ({e.reason}). Is the proxy running?") from e


def resolve_models(base: str, mk: str, group: str, override: str | None) -> list[str]:
    if override:
        return [m.strip() for m in override.split(",") if m.strip()]
    prefix = GROUP_PREFIXES[group]
    data = api(base, mk, "GET", "/v1/models")
    models = [m["id"] for m in data.get("data", []) if m["id"].startswith(prefix)]
    if not models:
        sys.exit(f"ERROR: proxy lists no models with prefix '{prefix}' for group '{group}'.")
    return models


def user_state(base: str, mk: str, user_id: str, alias: str) -> tuple[bool, list[str], str | None]:
    """Returns (user_exists, user_models, hashed_token_of_aliased_key_or_None)."""
    try:
        info = api(base, mk, "GET", f"/user/info?user_id={user_id}")
    except SystemExit:
        return False, [], None
    user_info = info.get("user_info") or {}
    exists = bool(user_info)
    models = list(user_info.get("models") or [])
    token = None
    for k in info.get("keys") or []:
        if k.get("key_alias") == alias:
            token = k.get("token")
            break
    return exists, models, token


def list_keys(base: str, mk: str, group_filter: str | None) -> None:
    """Print all manage-budget-key keys (alias '<group>-budget-<user>') as a table.
    Secrets are never stored, so only the masked key_name is shown."""
    data = api(base, mk, "GET", "/key/list?return_full_object=true&size=100")
    secrets = load_secrets()
    rows = []
    user_cache: dict[str, dict] = {}
    for k in data.get("keys") or []:
        alias = k.get("key_alias") or ""
        if "-budget-" not in alias:
            continue
        group = alias.split("-budget-", 1)[0]
        if group_filter and group != group_filter:
            continue
        user_id = k.get("user_id") or ""
        if user_id and user_id not in user_cache:
            try:
                user_cache[user_id] = (api(base, mk, "GET", f"/user/info?user_id={user_id}").get("user_info") or {})
            except SystemExit:
                user_cache[user_id] = {}
        u = user_cache.get(user_id, {})
        rows.append({
            "user": user_id,
            "group": group,
            "models": len(k.get("models") or []),
            "win_cap": k.get("max_budget"),
            "win_dur": k.get("budget_duration") or "",
            "win_spend": k.get("spend") or 0,
            "wk_cap": u.get("max_budget"),
            "wk_dur": u.get("budget_duration") or "",
            "wk_spend": u.get("spend") or 0,
            "key": secrets.get(alias) or (k.get("key_name") or "") + " (masked)",
            "alias": alias,
        })

    if not rows:
        print("No managed budget keys found." + (f" (group={group_filter})" if group_filter else ""))
        return

    def money(v) -> str:
        if v is None:
            return "-"
        if v != 0 and abs(v) < 0.01:
            return f"${v:.4f}"
        return f"${v:.2f}"

    rows.sort(key=lambda r: (r["group"], r["user"]))
    hdr = ["USER", "GROUP", "MODELS", "WINDOW", "WIN SPENT", "WEEKLY", "WK SPENT", "KEY", "ALIAS"]
    table = [hdr] + [[
        r["user"], r["group"], str(r["models"]),
        f'{money(r["win_cap"])}/{r["win_dur"]}', money(r["win_spend"]),
        f'{money(r["wk_cap"])}/{r["wk_dur"]}', money(r["wk_spend"]),
        r["key"], r["alias"],
    ] for r in rows]
    widths = [max(len(row[i]) for row in table) for i in range(len(hdr))]
    for ri, row in enumerate(table):
        print("  ".join(cell.ljust(widths[i]) for i, cell in enumerate(row)))
        if ri == 0:
            print("  ".join("-" * widths[i] for i in range(len(hdr))))
    print(f"\nNote: full sk-... values come from the local store {SECRETS_FILE.name} (git-ignored),")
    print("populated at --create time. Entries marked '(masked)' predate that store or were")
    print("created elsewhere — litellm keeps only a hash, so those secrets can't be recovered.")


def main() -> None:
    p = argparse.ArgumentParser(add_help=True, description=__doc__.splitlines()[2])
    mode = p.add_mutually_exclusive_group()
    mode.add_argument("--create", action="store_true", help="rotate: revoke old key, mint a new one (default)")
    mode.add_argument("--update", action="store_true", help="edit budgets in place; key value and spend preserved")
    mode.add_argument("--list", action="store_true", help="list all managed budget keys as a table (no secrets)")
    p.add_argument("--group", choices=sorted(GROUP_PREFIXES), default="deepseek", help="model group (default: deepseek)")
    p.add_argument("--models", help="comma-separated explicit model list (overrides --group's list)")
    p.add_argument("user_id", nargs="?", default="haitao")
    p.add_argument("weekly_usd", nargs="?", type=float, default=20.0)
    p.add_argument("window_usd", nargs="?", type=float, default=3.0)
    p.add_argument("window", nargs="?", default="5h")
    args = p.parse_args()

    base = os.environ.get("BASE", "http://localhost:4000")
    weekly_duration = os.environ.get("WEEKLY_DURATION", "1w")
    mk = load_master_key()

    if args.list:
        list_keys(base, mk, args.group if any(a in sys.argv for a in ("--group",)) else None)
        return

    alias = f"{args.group}-budget-{args.user_id}"
    models = resolve_models(base, mk, args.group, args.models)
    exists, user_models, token_hash = user_state(base, mk, args.user_id, alias)
    # Union with the user's current model access so granting one group doesn't
    # revoke another (the weekly cap stays a per-user aggregate across groups).
    merged_models = sorted(set(user_models) | set(models))

    user_body = {
        "user_id": args.user_id,
        "max_budget": args.weekly_usd,
        "budget_duration": weekly_duration,
        "models": merged_models,
    }

    if args.update:
        if not token_hash:
            sys.exit(f"ERROR: no existing key with alias '{alias}'. Run with --create first.")
        print(f"==> [update] Weekly cap -> ${args.weekly_usd:g} / {weekly_duration} (user-level)")
        api(base, mk, "POST", "/user/update", user_body)
        print(f"==> [update] {args.window} cap -> ${args.window_usd:g} (key-level, in place; spend kept)")
        # Identify the key by hashed token: the server only re-hashes values
        # starting with 'sk-', so a stored hash passes through unchanged.
        api(base, mk, "POST", "/key/update",
            {"key": token_hash, "max_budget": args.window_usd,
             "budget_duration": args.window, "models": models})
        print(f"""
✅ Budgets updated in place (key value unchanged).

  user_id : {args.user_id}
  weekly  : ${args.weekly_usd:g} / {weekly_duration}   (user-level)
  window  : ${args.window_usd:g} / {args.window}          (key-level)
  group   : {args.group}  ({len(models)} models)
  alias   : {alias}

The existing key keeps working — no need to re-paste it into the extension.
Accumulated spend and the current reset window are preserved.""")
        return

    # ------------------------------------------------------------- CREATE
    print(f"==> [create] Ensuring user '{args.user_id}' with weekly cap ${args.weekly_usd:g} / {weekly_duration}")
    if exists:
        api(base, mk, "POST", "/user/update", user_body)
    else:
        # auto_create_key=false: don't mint an extra unbudgeted default key.
        api(base, mk, "POST", "/user/new", {**user_body, "auto_create_key": False})

    print(f"==> [create] Revoking any existing key with alias '{alias}'")
    if token_hash:
        deleted = api(base, mk, "POST", "/key/delete", {"key_aliases": [alias]})
        print(f"    revoked {len(deleted.get('deleted_keys') or [])} old key(s)")
    else:
        print("    revoked 0 old key(s)")

    print(f"==> [create] Generating new key with {args.window} cap ${args.window_usd:g} ({args.group})")
    resp = api(base, mk, "POST", "/key/generate",
               {"user_id": args.user_id, "key_alias": alias,
                "max_budget": args.window_usd, "budget_duration": args.window,
                "models": models})
    key = resp.get("key")
    if not key:
        sys.exit(f"ERROR: no key in response: {json.dumps(resp)[:400]}")
    save_secret(alias, key)  # so --list can show the full secret later

    print(f"""
✅ Key created.

  user_id : {args.user_id}
  weekly  : ${args.weekly_usd:g} / {weekly_duration}   (user-level)
  window  : ${args.window_usd:g} / {args.window}          (key-level)
  group   : {args.group}  ({len(models)} models)
  alias   : {alias}
  key     : {key}

Use this key as the API key in the VS Code extension (Base URL: {base}).
It stops serving when EITHER the weekly or the {args.window} budget is exhausted.

NOTE: --create ROTATES — any prior '{alias}' key was revoked, so re-paste this
new value. To change only budgets without rotating (keeps spend + window):
  ./litellm/manage-budget-key.py --update --group {args.group} {args.user_id} <weekly> <window_usd>""")


if __name__ == "__main__":
    main()

File: litellm/manage-budget-key.ts
#!/usr/bin/env bun
/**
 * manage-budget-key.ts
 * --------------------
 * Manage a model-group-scoped LiteLLM virtual key with Claude-Code-style
 * layered budgets:
 *
 *   - a weekly cap   (enforced on the user)
 *   - a window cap   (default 5h, enforced on the key)
 *
 * The key blocks when EITHER limit is hit.
 *
 * Modes:
 *   --create  (default)  Revoke any existing key for the user+group and mint
 *                        a NEW one. New sk-... value, window spend restarts at $0.
 *   --update             Change the budgets on the EXISTING key in place. Same
 *                        sk-... value; accumulated spend and reset window kept.
 *   --list               Show all managed budget keys as a table (budgets +
 *                        spend; only a masked key is shown). --group filters.
 *
 * Model groups (--group, default "deepseek"):
 *   deepseek  -> models from the proxy whose id starts with "deepseek-"
 *   copilot   -> models whose id starts with "github_copilot/"
 *   (--models a,b,c overrides the list explicitly)
 *
 * The group also names the key alias: "<group>-budget-<user_id>", which is how
 * one budgeted key per user per group is found and rotated without its
 * plaintext. The weekly cap lives on the USER, so it aggregates across all of
 * that user's keys/groups; user-level model access is the union of all groups.
 *
 * Usage:
 *   ./litellm/manage-budget-key.ts [--create|--update] [--group G] [--models a,b]
 *                                  [user_id] [weekly_usd] [window_usd] [window]
 *
 * Defaults: user_id=haitao  weekly_usd=20  window_usd=3  window=5h
 *
 * Requirements: proxy running (BASE, default http://localhost:4000);
 * LITELLM_MASTER_KEY set (sourced from .env at repo root if present).
 */

import { readFileSync, writeFileSync, existsSync, chmodSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const GROUP_PREFIXES: Record<string, string> = {
	deepseek: "deepseek-",
	copilot: "github_copilot/",
};
// Local plaintext store of minted secrets, keyed by alias. litellm only stores a
// hash, so this is the only way --list can show full keys. Git-ignored, chmod 600.
const SECRETS_FILE = resolve(ROOT, ".litellm-budget-keys.json");

function loadSecrets(): Record<string, string> {
	try {
		return JSON.parse(readFileSync(SECRETS_FILE, "utf8"));
	} catch {
		return {};
	}
}

function saveSecret(alias: string, key: string): void {
	const secrets = loadSecrets();
	secrets[alias] = key;
	writeFileSync(SECRETS_FILE, JSON.stringify(secrets, null, 2));
	try {
		chmodSync(SECRETS_FILE, 0o600);
	} catch {
		/* best effort */
	}
}

function fail(msg: string): never {
	console.error(msg);
	process.exit(1);
}

// ---------------------------------------------------------------- arg parse
interface Args {
	mode: "create" | "update" | "list";
	group: string;
	groupExplicit: boolean;
	models?: string;
	userId: string;
	weeklyUsd: number;
	windowUsd: number;
	window: string;
}

function parseArgs(argv: string[]): Args {
	const a: Args = {
		mode: "create",
		group: "deepseek",
		groupExplicit: false,
		userId: "haitao",
		weeklyUsd: 20,
		windowUsd: 3,
		window: "5h",
	};
	const pos: string[] = [];
	for (let i = 0; i < argv.length; i++) {
		const t = argv[i];
		if (t === "--create") a.mode = "create";
		else if (t === "--update") a.mode = "update";
		else if (t === "--list") a.mode = "list";
		else if (t === "--group") {
			const v = argv[++i];
			if (!v || !(v in GROUP_PREFIXES))
				fail(`ERROR: --group must be one of: ${Object.keys(GROUP_PREFIXES).join(", ")}`);
			a.group = v;
			a.groupExplicit = true;
		} else if (t === "--models") {
			a.models = argv[++i] ?? fail("ERROR: --models needs a comma-separated list");
		} else if (t === "--help" || t === "-h") {
			console.log(
				"Usage: manage-budget-key.ts [--create|--update|--list] [--group deepseek|copilot] [--models a,b] [user_id] [weekly_usd] [window_usd] [window]"
			);
			process.exit(0);
		} else if (t.startsWith("--")) fail(`ERROR: unknown flag '${t}' (expected --create, --update or --list)`);
		else pos.push(t);
	}
	if (pos[0]) a.userId = pos[0];
	if (pos[1]) a.weeklyUsd = Number(pos[1]);
	if (pos[2]) a.windowUsd = Number(pos[2]);
	if (pos[3]) a.window = pos[3];
	if (!Number.isFinite(a.weeklyUsd) || !Number.isFinite(a.windowUsd)) fail("ERROR: budgets must be numbers");
	return a;
}

// ---------------------------------------------------------------- env / api
function loadMasterKey(): string {
	if (!process.env.LITELLM_MASTER_KEY) {
		const envFile = resolve(ROOT, ".env");
		if (existsSync(envFile)) {
			for (const line of readFileSync(envFile, "utf8").split("\n")) {
				const m = line.trim().match(/^LITELLM_MASTER_KEY=(.*)$/);
				if (m) {
					process.env.LITELLM_MASTER_KEY = m[1];
					break;
				}
			}
		}
	}
	return process.env.LITELLM_MASTER_KEY ?? fail("ERROR: LITELLM_MASTER_KEY is not set (export it or put it in .env).");
}

const BASE = process.env.BASE ?? "http://localhost:4000";
const WEEKLY_DURATION = process.env.WEEKLY_DURATION ?? "1w";
const MK = loadMasterKey();

async function api(method: string, path: string, body?: unknown, tolerate404 = false): Promise<any> {
	let res: Response;
	try {
		res = await fetch(BASE + path, {
			method,
			headers: { Authorization: `Bearer ${MK}`, "Content-Type": "application/json" },
			body: body === undefined ? undefined : JSON.stringify(body),
		});
	} catch (e) {
		fail(`ERROR: cannot reach ${BASE} (${e}). Is the proxy running?`);
	}
	if (res.status === 404 && tolerate404) return null;
	if (!res.ok) fail(`ERROR: ${method} ${path} -> HTTP ${res.status}: ${(await res.text()).slice(0, 400)}`);
	return res.json();
}

// ---------------------------------------------------------------- helpers
async function resolveModels(group: string, override?: string): Promise<string[]> {
	if (override)
		return override
			.split(",")
			.map((s) => s.trim())
			.filter(Boolean);
	const prefix = GROUP_PREFIXES[group];
	const data = await api("GET", "/v1/models");
	const models = (data.data ?? []).map((m: { id: string }) => m.id).filter((id: string) => id.startsWith(prefix));
	if (models.length === 0) fail(`ERROR: proxy lists no models with prefix '${prefix}' for group '${group}'.`);
	return models;
}

interface UserState {
	exists: boolean;
	userModels: string[];
	tokenHash?: string;
}

async function userState(userId: string, alias: string): Promise<UserState> {
	// 404 = user doesn't exist yet — that's a normal first-run state.
	const info = await api("GET", `/user/info?user_id=${encodeURIComponent(userId)}`, undefined, true);
	if (info === null) return { exists: false, userModels: [] };
	const userInfo = info.user_info ?? {};
	const keys: Array<{ key_alias?: string; token?: string }> = info.keys ?? [];
	return {
		exists: Object.keys(userInfo).length > 0,
		userModels: userInfo.models ?? [],
		tokenHash: keys.find((k) => k.key_alias === alias)?.token,
	};
}

// ---------------------------------------------------------------- list
function money(v: number | null | undefined): string {
	if (v === null || v === undefined) return "-";
	if (v !== 0 && Math.abs(v) < 0.01) return `$${v.toFixed(4)}`;
	return `$${v.toFixed(2)}`;
}

async function listKeys(groupFilter?: string): Promise<void> {
	const data = await api("GET", "/key/list?return_full_object=true&size=100");
	const secrets = loadSecrets();
	const userCache: Record<string, any> = {};
	const rows: string[][] = [];
	for (const k of data.keys ?? []) {
		const alias: string = k.key_alias ?? "";
		if (!alias.includes("-budget-")) continue;
		const group = alias.split("-budget-")[0];
		if (groupFilter && group !== groupFilter) continue;
		const userId: string = k.user_id ?? "";
		if (userId && !(userId in userCache)) {
			const info = await api("GET", `/user/info?user_id=${encodeURIComponent(userId)}`, undefined, true);
			userCache[userId] = info?.user_info ?? {};
		}
		const u = userCache[userId] ?? {};
		rows.push([
			userId,
			group,
			String((k.models ?? []).length),
			`${money(k.max_budget)}/${k.budget_duration ?? ""}`,
			money(k.spend ?? 0),
			`${money(u.max_budget)}/${u.budget_duration ?? ""}`,
			money(u.spend ?? 0),
			secrets[alias] ?? `${k.key_name ?? ""} (masked)`,
			alias,
		]);
	}
	if (rows.length === 0) {
		console.log("No managed budget keys found." + (groupFilter ? ` (group=${groupFilter})` : ""));
		return;
	}
	rows.sort((a, b) => a[1].localeCompare(b[1]) || a[0].localeCompare(b[0]));
	const hdr = ["USER", "GROUP", "MODELS", "WINDOW", "WIN SPENT", "WEEKLY", "WK SPENT", "KEY", "ALIAS"];
	const table = [hdr, ...rows];
	const widths = hdr.map((_, i) => Math.max(...table.map((r) => r[i].length)));
	table.forEach((r, ri) => {
		console.log(r.map((c, i) => c.padEnd(widths[i])).join("  "));
		if (ri === 0) console.log(widths.map((w) => "-".repeat(w)).join("  "));
	});
	console.log(`\nNote: full sk-... values come from the local store ${"."}litellm-budget-keys.json (git-ignored),`);
	console.log("populated at --create time. Entries marked '(masked)' predate that store or were");
	console.log("created elsewhere — litellm keeps only a hash, so those secrets can't be recovered.");
}

// ---------------------------------------------------------------- main
const args = parseArgs(process.argv.slice(2));

if (args.mode === "list") {
	await listKeys(args.groupExplicit ? args.group : undefined);
	process.exit(0);
}

const alias = `${args.group}-budget-${args.userId}`;
const models = await resolveModels(args.group, args.models);
const { exists, userModels, tokenHash } = await userState(args.userId, alias);
// Union with current access so granting one group doesn't revoke another.
const mergedModels = [...new Set([...userModels, ...models])].sort();

const userBody = {
	user_id: args.userId,
	max_budget: args.weeklyUsd,
	budget_duration: WEEKLY_DURATION,
	models: mergedModels,
};

if (args.mode === "update") {
	if (!tokenHash) fail(`ERROR: no existing key with alias '${alias}'. Run with --create first.`);
	console.log(`==> [update] Weekly cap -> $${args.weeklyUsd} / ${WEEKLY_DURATION} (user-level)`);
	await api("POST", "/user/update", userBody);
	console.log(`==> [update] ${args.window} cap -> $${args.windowUsd} (key-level, in place; spend kept)`);
	// Identify the key by hashed token: the server only re-hashes values
	// starting with 'sk-', so a stored hash passes through unchanged.
	await api("POST", "/key/update", {
		key: tokenHash,
		max_budget: args.windowUsd,
		budget_duration: args.window,
		models,
	});
	console.log(`
✅ Budgets updated in place (key value unchanged).

  user_id : ${args.userId}
  weekly  : $${args.weeklyUsd} / ${WEEKLY_DURATION}   (user-level)
  window  : $${args.windowUsd} / ${args.window}          (key-level)
  group   : ${args.group}  (${models.length} models)
  alias   : ${alias}

The existing key keeps working — no need to re-paste it into the extension.
Accumulated spend and the current reset window are preserved.`);
	process.exit(0);
}

// --------------------------------------------------------------- create
console.log(`==> [create] Ensuring user '${args.userId}' with weekly cap $${args.weeklyUsd} / ${WEEKLY_DURATION}`);
if (exists) {
	await api("POST", "/user/update", userBody);
} else {
	// auto_create_key=false: don't mint an extra unbudgeted default key.
	await api("POST", "/user/new", { ...userBody, auto_create_key: false });
}

console.log(`==> [create] Revoking any existing key with alias '${alias}'`);
if (tokenHash) {
	const deleted = await api("POST", "/key/delete", { key_aliases: [alias] });
	console.log(`    revoked ${(deleted.deleted_keys ?? []).length} old key(s)`);
} else {
	console.log("    revoked 0 old key(s)");
}

console.log(`==> [create] Generating new key with ${args.window} cap $${args.windowUsd} (${args.group})`);
const resp = await api("POST", "/key/generate", {
	user_id: args.userId,
	key_alias: alias,
	max_budget: args.windowUsd,
	budget_duration: args.window,
	models,
});
if (!resp.key) fail(`ERROR: no key in response: ${JSON.stringify(resp).slice(0, 400)}`);
saveSecret(alias, resp.key); // so --list can show the full secret later

console.log(`
✅ Key created.

  user_id : ${args.userId}
  weekly  : $${args.weeklyUsd} / ${WEEKLY_DURATION}   (user-level)
  window  : $${args.windowUsd} / ${args.window}          (key-level)
  group   : ${args.group}  (${models.length} models)
  alias   : ${alias}
  key     : ${resp.key}

Use this key as the API key in the VS Code extension (Base URL: ${BASE}).
It stops serving when EITHER the weekly or the ${args.window} budget is exhausted.

NOTE: --create ROTATES — any prior '${alias}' key was revoked, so re-paste this
new value. To change only budgets without rotating (keeps spend + window):
  ./litellm/manage-budget-key.ts --update --group ${args.group} ${args.userId} <weekly> <window_usd>`);
